### PR TITLE
Speedups to HGCal clustering

### PR DESCRIPTION
**Pull request by lgray**
Addressing comments made at the end of: https://github.com/cms-sw/cmssw/pull/16909

This PR significantly decreases the number of calls to sqrt() in tight loops and uses vdt where useful in the HGCal layer and multiclusters.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please abort

better start with a new IB present
This PR is supposedly technical

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
this is still too early
there is no working baseline with multicluster PR included.

**Comment by lgray** [comment]
please abort

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
Unit test failure is not related to this PR. The unit test is fixed in: #17010

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

unit test fixed.

File: RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc [hunk 1]
```diff
@@ -83,14 +85,16 @@ math::XYZPoint ClusterTools::getMultiClusterPosition(const reco::HGCalMultiClust
   for( const auto& ptr : clu.clusters() ) {   
     const double x = ptr->x();    
     const double y = ptr->y();
-    const float point_r = std::sqrt(x*x + y*y);    
+    const double point_r2 = (x*x + y*y);
```
Comment (lgray) [inline RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc]: everything else is already in double, repeated conversion from double to float costs time.

Comment (lgray) [inline RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc]: and this is in a tight loop

Comment (slava77) [inline RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc]: the reason for x,y,z and energy to be in double is debatable on the storage side.
there is nothing else inside this loop that requires double.

OK for now

File: RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc [hunk 2]
```diff
@@ -172,9 +172,13 @@ math::XYZPoint HGCalImagingAlgo::calculatePosition(std::vector<KDNode> &v){
 } 
 
 double HGCalImagingAlgo::distance(const Hexel &pt1, const Hexel &pt2){
+  return std::sqrt(distance2(pt1,pt2));
+}
+
+double HGCalImagingAlgo::distance2(const Hexel &pt1, const Hexel &pt2){
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc]: can this be inlined?

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16997 f921ac6
- code optimization to speedup hgcalLayerClusters  HGCalClusterTestProducer
- jenkins tests pass and comparisons with baseline show no differences
- local tests with ttbar with D4Timing PU200, running just RECO:hgcalLocalRecoSequence show moderate speedup: hgcalLayerClusters time changes from 31.6 s to 25.4 s per event.

**Comment by davidlange6** [comment]
+1

**Pull request merged**